### ISSUE TITLE
Existentials

### ISSUE THREAD

**Issue by georgefst** [dhall-lang/dhall-haskell#1906]
My Haskell library uses a parameterised `Layout a` type, which can be parsed from a Dhall expression. Different backends will use a different `a`. This allows them to deal with different sorts of metadata.

But I have a demo implementation which doesn't care what `a` is. i.e. it exposes a function:

```haskell
f :: forall a. Layout a -> b
```

Is there any way I can accept Dhall expressions with *any* `a`, to feed to `f`? Perhaps a `FromDhall` instance for:

```haskell
data AnyLayout where
    AnyLayout :: Layout a -> AnyLayout
```

Failing that, it would be convenient if I could have a special `Any` type, with a `FromDhall` instance that just accepts anything. Then I could use `Layout Any`. But that doesn't seem possible. This is related to #1521.

**Comment by Gabriel439** [comment]
@georgefst: Could you provide a simplified version of the `Layout` type?  If I know how the `a` within `Layout a` is used then I might be able to suggest a solution

**Comment by georgefst** [comment]
It corresponds to one field in a deeply nested record. A simplified version would be:

```haskell
data Layout a = Layout
    { elements :: [Element a]
    }

data Element a = Element
    { position :: Int
    , metadata :: a
    }
```

**Comment by MonoidMusician** [comment]
If you’re just ignoring the information, why not map to be the unit type `{}`, so you’re always just dealing with `Layout {}`?

Otherwise, an encoding for existentials would be: `forall (r : Type) -> (forall (a : Type) -> f a -> r) -> r`. That’s a mouthful and I don’t think we have auto decoding for that yet.

**Comment by georgefst** [comment]
> why not map to be the unit type {}

That's what I'm currently doing. I have a function:

```dhall
void : ∀(a : Type) → Layout a → Layout {}
```

The annoying thing is the lack of type inference - I have to manually pass the type argument, so I can't just 'blindly' accept *any* `Layout a`.

Additionally, defining `void` is slightly tedious compared to just `deriving Functor` - it would be nice to shift more of the logic to the Haskell side.

**Comment by Gabriel439** [comment]
@georgefst: I believe this is essentially the same issue as https://github.com/dhall-lang/dhall-haskell/issues/1521.  Specifically, both require the ability to load an arbitrary type of value somewhere within the schema without failing at type-checking time

**Comment by georgefst** [comment]
@Gabriel439 

Yes, having now read that thread more fully, it does seem that the kinds of solutions discussed there would also work for me.

I particularly like the idea of a `FromDhall (Expr a s)` instance, which is not one that I'd previously considered. Is that likely to be added to the library?

**Comment by 46511dedi** [comment]
,😊

**Comment by Gabriel439** [comment]
@georgefst: The main blocker is that we'd need some way to disable type-checking for an expression that contained an `Expr s a` anywhere within its type.  Essentially when you decode a Dhall value into a Haskell value it first checks that the value is the correct type and then decodes it.  This should be possible, but it requires a bit of work to avoid API disruption and also to not lead to surprising behavior for end users

**Comment by georgefst** [comment]
@Gabriel439 Fair enough. I don't have enough knowledge of the internals to know how disruptive that would be.

I'm happy for this to be closed in favour of #1521, so long as there's somewhere to track progress.

**Comment by Gabriel439** [comment]
Alright, then I'll close this in favor of #1521
